rustdoc: show inner enum and struct in type definition for concrete type

Urgau opened the pull request:
This PR implements the [Display enum variants for generic enum in type def page](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/Display.20enum.20variants.20for.20generic.20enum.20in.20type.20def.20page) #rustdoc/zulip proposal.

This proposal comes from looking at [`TyKind`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/sty/type.TyKind.html) typedef from the compiler. On that page, the documentation is able to show the layout for each variant, but not the variants themselves. This proposal suggests showing the fields and variants for those "concrete type". This would mean that instead of having many unresolved generics, like in `IrTyKind`:
```rust
    Array(I::Ty, I::Const),
    Slice(I::Ty),
    RawPtr(I::TypeAndMut),
    Ref(I::Region, I::Ty, I::Mutability),
    FnDef(I::DefId, I::GenericArgsRef),
```
those would be resolved with direct links to the proper types in the `TyKind` typedef page:
```rust
    Array(Ty<'tcx>, Const<'tcx>),
    Slice(Ty<'tcx>),
    RawPtr(TypeAndMut<'tcx>),
    Ref(Region<'tcx>, Ty<'tcx>, Mutability<'tcx>),
    FnDef(DefId<'tcx>, GenericArgsRef<'tcx>),
```
Saving both time and confusion.

-----

<details>

<summary>Old description</summary>

I've chosen to add the enums and structs under the "Show Aliased Type" details, as well as showing the variants and fields under the usual "Variants" and "Fields" sections. ~~*under new the `Inner Variants` and `Inner Fields` sections (except for their names, they are identical to the one found in the enum, struct and union pages). Those sections are complementary and do not replace anything else.*~~

This PR proposes the following condition for showing the aliased type (basically, has the aliased type some generics that are all of them resolved):
 - the typedef does NOT have any generics (modulo lifetimes)
 - AND the aliased type has some generics

</details>

### Examples

```rust
pub enum IrTyKind<'a, I: Interner> {
    /// Doc comment for AdtKind
    AdtKind(&'a I::Adt),
    /// and another one for TyKind
    TyKind(I::Adt, I::Ty),
    // no comment
    StructKind { a: I::Adt, },
}

pub type TyKind<'a> = IrTyKind<'a, TyCtxt>;
```
![TyKind](https://github.com/rust-lang/rust/assets/3616612/13307679-6d48-40d6-ad50-6db0b7f36ac7)

<details>
<summary>Old</summary>

![image](https://github.com/rust-lang/rust/assets/3616612/4147c049-d056-42d4-8a01-d43ebe747308)

![TyKind](https://user-images.githubusercontent.com/3616612/260988247-34831aa9-470d-4286-ad9f-3e8002153a92.png)

![TyKind](https://github.com/rust-lang/rust/assets/3616612/62381bb3-fa0f-4b05-926d-77759cf9115a)

</details>

```rust
pub struct One<T> {
    pub val: T,
    #[doc(hidden)]
    pub inner_tag: u64,
    __hidden: T,
}

/// `One` with `u64` as payload
pub type OneU64 = One<u64>;
```
![OneU64](https://github.com/rust-lang/rust/assets/3616612/d551b474-ce88-4f8c-bc94-5c88aba51424)

<details>
<summary>Old</summary>

![image](https://github.com/rust-lang/rust/assets/3616612/1a3f53c0-17bf-4aa7-894d-3fedc15b33da)

![OneU64](https://github.com/rust-lang/rust/assets/3616612/7b124a5b-e287-4efb-b9ca-fdcd1cdeeba8)

![OneU64](https://github.com/rust-lang/rust/assets/3616612/ddd962be-4f76-4ecd-81bd-531f3dd23832)

</details>

r? @GuillaumeGomez

Urgau committed:
rustdoc: show inner enum and struct in type definition for concrete type

GuillaumeGomez commented:
Just having "inner fields" isn't great. Maybe we could show the aliased type definition in a "show aliased type" thing like we have for "show all" when there are too any fields/variants?

Urgau committed:
rustdoc: extract logic for printing enum fields and struct variants

Urgau commented:
I've discussed with @GuillaumeGomez on Zulip about the UI layout and we concluded that it should also display the enum/struct code definition. So I've added it.

![image](https://github.com/rust-lang/rust/assets/3616612/34831aa9-470d-4286-ad9f-3e8002153a92)

GuillaumeGomez commented on src/librustdoc/clean/mod.rs:
```diff
@@ -955,6 +956,42 @@ fn clean_ty_generics<'tcx>(
 
+fn clean_ty_typedef_inner_type<'tcx>(ty: Ty<'tcx>, cx: &mut DocContext<'tcx>) -> TypedefInnerType {
+    if let ty::Adt(adt_def, args) = ty.kind() {
+        if !adt_def.is_struct() {
```
Please use [is_enum](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.AdtDef.html#method.is_enum) instead (`is_struct` doesn't return `true` for unions as far as I can see).

GuillaumeGomez commented on src/librustdoc/clean/types.rs:
```diff
@@ -2229,10 +2229,25 @@ pub(crate) struct PathSegment {
+        /* has_stripped_entries */ bool,
+        /* is_non_exhaustive */ bool,
+    ),
+    Struct(Option<CtorKind>, Vec<Item>, /* has_stripped_entries */ bool),
```
There is also union types. So either you rename this variant into "StructLike" (or something along the line) or you add a `Union` variant.

Urgau replied:
Added a `Union` variant.

GuillaumeGomez commented on src/librustdoc/clean/types.rs:
```diff
@@ -2229,10 +2229,25 @@ pub(crate) struct PathSegment {
     pub(crate) generics: Generics,
+    /// Inner variants/fields for fully resolved type def, ie `type TyKind = IrTyKind<Adt, Ty>`,
+    /// to be shown directly on the typedef page.
+    pub(crate) inner_type: TypedefInnerType,
```
Could you use an `Option` instead of having `TypedefInnerType::None` please?

GuillaumeGomez commented on src/librustdoc/html/render/print_item.rs:
```diff
@@ -1238,6 +1238,72 @@ fn item_typedef(w: &mut Buffer, cx: &mut Context<'_>, it: &clean::Item, t: &clea
+        && t.type_.generics().map(|generics| !generics.is_empty()).unwrap_or(false);
+
+    if should_print_inner_variants {
+        fn toggle_open(mut w: impl fmt::Write) {
```
Rather than this, could you instead create a `toggle_wrapper` which will call this write, then call the closure to generate the "inner code" then call `toggle_close` at the end? Like that, impossible to forget a close and simpler to follow.

GuillaumeGomez commented on src/librustdoc/html/render/print_item.rs:
```diff
@@ -1342,148 +1408,179 @@ fn item_enum(w: &mut Buffer, cx: &mut Context<'_>, it: &clean::Item, e: &clean::
+        write!(
+            w,
+            "<section id=\"{id}\" class=\"variant\">\
                     <a href=\"#{id}\" class=\"anchor\">§</a>",
```
You forgot to change the indent on this line.

GuillaumeGomez commented on src/librustdoc/html/render/print_item.rs:
```diff
@@ -1342,148 +1408,179 @@ fn item_enum(w: &mut Buffer, cx: &mut Context<'_>, it: &clean::Item, e: &clean::
+    let tcx = cx.tcx();
+    write!(
+        w,
+        "<h2 id=\"variants\" class=\"variants small-section-header\">\
+                {prefix}Variants{}<a href=\"#variants\" class=\"anchor\">§</a>\
             </h2>\
             {}\
             <div class=\"variants\">",
```
You forgot to change the indent on these lines.

GuillaumeGomez commented on src/librustdoc/html/render/print_item.rs:
```diff
@@ -1342,148 +1408,179 @@ fn item_enum(w: &mut Buffer, cx: &mut Context<'_>, it: &clean::Item, e: &clean::
+                cx.derive_id(format!("{}.{}.fields", ItemType::Variant, variant.name.unwrap()));
+            write!(
+                w,
+                "<div class=\"sub-variant\" id=\"{variant_id}\">\
                         <h4>{heading}</h4>\
                         {}",
```
Indent.

GuillaumeGomez commented on src/librustdoc/html/render/print_item.rs:
```diff
@@ -1342,148 +1408,179 @@ fn item_enum(w: &mut Buffer, cx: &mut Context<'_>, it: &clean::Item, e: &clean::
+            match *v.kind {
+                // FIXME(#101337): Show discriminant
+                clean::VariantItem(ref var) => match var.kind {
+                    clean::VariantKind::CLike => write!(w, "{}", name),
```
```suggestion
                    clean::VariantKind::CLike => w.write_str(name.as_str()),
```

GuillaumeGomez commented:
Looks really nice! I'm a big fan of this change considering how often we encounter it in rustc's docs. Once we're done with this review round, I'll start an FCP as it is a UI change.

fmease commented on src/librustdoc/clean/types.rs:
```diff
@@ -2229,10 +2229,24 @@ pub(crate) struct PathSegment {
 
+#[derive(Clone, Debug)]
+pub(crate) enum TypedefInnerType {
+    Enum(
+        IndexVec<VariantIdx, Item>,
+        /* has_stripped_entries */ bool,
+        /* is_non_exhaustive */ bool,
+    ),
+    Union(Vec<Item>, /* has_stripped_fields */ bool),
+    Struct(Option<CtorKind>, Vec<Item>, /* has_stripped_fields */ bool),
```
```suggestion
    Enum {
        variants: IndexVec<VariantIdx, Item>,
        has_stripped_entries: bool,
        is_non_exhaustive: bool,
    },
    Union {
        fields: Vec<Item>,
        has_stripped_fields: bool,
    },
    Struct {
        kind: Option<CtorKind>,
        fields: Vec<Item>,
        has_stripped_fields: bool,
     },
```

Urgau replied:
Thanks, that's indeed much better.

fmease commented on src/librustdoc/html/render/print_item.rs:
```diff
@@ -1238,6 +1238,101 @@ fn item_typedef(w: &mut Buffer, cx: &mut Context<'_>, it: &clean::Item, t: &clea
+        .iter()
+        .all(|param| matches!(param.kind, clean::GenericParamDefKind::Lifetime { .. }))
+        && t.generics.where_predicates.is_empty()
+        && t.type_.generics().map(|generics| !generics.is_empty()).unwrap_or(false);
```
```suggestion
        && t.type_.generics().map_or(false, |generics| !generics.is_empty());
```


```suggestion
        && t.type_.generics().is_some_and(|generics| !generics.is_empty());
```

fmease commented on src/librustdoc/html/render/print_item.rs:
```diff
@@ -1342,148 +1437,180 @@ fn item_enum(w: &mut Buffer, cx: &mut Context<'_>, it: &clean::Item, e: &clean::
+    has_stripped_entries: bool,
+    is_non_exhaustive: bool,
+) {
+    if !g.map(|g| print_where_clause_and_check(w, g, cx)).unwrap_or(false) {
```
Consider using `map_or` or `is_some_and`.

fmease reviewed (commented):
Some small style suggestions. I haven't looked at the changes from a semantic standpoint yet, I'm gonna do so earliest tomorrow.

Edit: Thanks a lot for working on this, this will help tremendously when browsing the rustc docs!

Urgau committed:
rustdoc: handle typedef inner type when doing cross-crate inlining

bors commented:
:umbrella: The latest upstream changes (presumably #114905) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fmease commented on src/librustdoc/clean/mod.rs:
```diff
@@ -2350,6 +2399,60 @@ pub(crate) fn clean_variant_def<'tcx>(variant: &ty::VariantDef, cx: &mut DocCont
+            variant
+                .fields
+                .iter()
+                .filter(|field| field.vis.is_public())
```
```suggestion
                .filter(|field| cx.render_options.document_private || field.vis.is_public())
```

fmease commented on src/librustdoc/clean/mod.rs:
```diff
@@ -2350,6 +2399,60 @@ pub(crate) fn clean_variant_def<'tcx>(variant: &ty::VariantDef, cx: &mut DocCont
+            fields: variant
+                .fields
+                .iter()
+                .filter(|field| field.vis.is_public())
```
```suggestion
                .filter(|field| cx.render_options.document_private || field.vis.is_public())
```

fmease reviewed (changes requested):
Have you checked if it handles `#[doc(hidden)]` correctly (with and without `--document-hidden-items -Zunstable-options`)? On fields and on variants?

Edit: I've just checked myself, it doesn't.

fmease commented:
The original motivation of this change is improving the UX for `rustc_middle::ty::Ty` (which heavily uses associated types). However at the moment, the implementation still falls short of it since we only instantiate the inner type but not normalize the result. We end up with (which doesn't help us):

![Screenshot 2023-08-18 at 15-29-05 TyKind in inner - Rust](https://github.com/rust-lang/rust/assets/14913065/5b9aa678-6219-45ea-9380-920ef23e805f)

Try doing `tcx.try_normalize_erasing_regions(instantiated_ty).unwrap_or(instantiated_ty)`.

Urgau committed:
rustdoc: bind typedef inner type items to the folding system

This let's us handle a multitude of things for free:
 - #[doc(hidden)]
 - private fields/variants
 - --document-private-items
 - --document-hidden-items

And correct in the process the determination of "has stripped items" by
doing the same logic done by other ones.

Urgau committed:
rustdoc: normalize all typedef inner types

Urgau commented on src/librustdoc/clean/mod.rs:
```diff
@@ -955,6 +956,46 @@ fn clean_ty_generics<'tcx>(
+            .map(|variant| clean_variant_def_with_args(variant, args, cx))
+            .collect();
+
+        let has_stripped_variants = adt_def.variants().len() != variants.len();
```
Hum... Yeah your right, it's always false, and it turns out was a symptom of a bigger issue around struct variants handling. Fixed!

Urgau commented:
> Have you checked if you handle `#[doc(hidden)]` correctly (with and without `--document-hidden-items -Zunstable-options`)? On fields and on variants?

Not before you asked, but I have now done it and fixed all the bugs related to it.

> The original motivation for this change is improving the UX for `rustc_middle::ty::TyKind` (which contains tons of associated types). However at the moment, the implementation still falls short in this exact use case since we only instantiate the inner type but don't normalize the result.

I didn't realize that, my bad. Fixed with your suggestion, thanks.

fmease commented on tests/rustdoc/typedef-inner-variants.rs:
```diff
@@ -0,0 +1,114 @@
+// @has 'inner_variants/enum.IrTyKind.html'
+pub enum IrTyKind<A, I: Interner> {
+    /// Doc comment for AdtKind
+    AdtKind(<I as Interner>::Adt),
+    /// and another one for TyKind
+    TyKind(<I as Interner>::Adt, <I as Interner>::Ty),
```
Strictly speaking, you don't need to fully qualify the associated type but it's fine.

```suggestion
    AdtKind(I::Adt),
    /// and another one for TyKind
    TyKind(I::Adt, I::Ty),
```

Urgau replied:
:+1: Changed for all, except for `Ty` in case it makes any difference :smile:.

fmease commented on src/librustdoc/clean/mod.rs:
```diff
@@ -2350,6 +2396,83 @@ pub(crate) fn clean_variant_def<'tcx>(variant: &ty::VariantDef, cx: &mut DocCont
+                    // normalize the type to only show concrete types
+                    // note: we do not use try_normalize_erasing_regions since we
+                    // do care about showing the regions
+                    let ty = infcx
+                        .at(&ObligationCause::dummy(), cx.param_env)
+                        .query_normalize(ty)
+                        .map(|normalized| normalized.value)
+                        .unwrap_or(ty);
```
:heart:

fmease commented on src/librustdoc/clean/mod.rs:
```diff
@@ -2350,6 +2396,83 @@ pub(crate) fn clean_variant_def<'tcx>(variant: &ty::VariantDef, cx: &mut DocCont
+        ty::VariantDiscr::Relative(_) => None,
+    };
+
+    use crate::rustc_trait_selection::infer::TyCtxtInferExt;
+    use crate::rustc_trait_selection::traits::query::normalize::QueryNormalizeExt;
```
```suggestion
    use rustc_trait_selection::infer::TyCtxtInferExt;
    use rustc_trait_selection::traits::query::normalize::QueryNormalizeExt;
```

fmease reviewed (approved):
Thanks, great improvements! I love that you now `fold` over the inner type instead of trying to manually strip items. Good call using `QueryNormalizer` instead of `try_normalize_erasing_regions` to respect regions.

Implementation now looks ready for an FCP (of course I might have missed some things)

@rustbot label A-rustdoc-ui

camelid commented:
I like this idea 👍. One small thing is I feel that the names "Inner Variants" and "Inner Fields" are not ideal. What if we just called them "Variants" and "Fields"? It'd be slightly confusing given that it's a type alias, but since to the user the alias behaves transparently like an enum or struct, it seems reasonable to me.

Urgau commented:
Sure. I had initially chosen to prefix those sections with "Inner" when it wasn't so clear that those fields and variants where coming from the aliased type, but since we now have the "Show Aliased Type" detail and the definition of the aliased type, it makes sense to remove the "Inner" prefix. Done.

GuillaumeGomez commented:
@rfcbot fcp merge

bors commented:
:umbrella: The latest upstream changes (presumably #115078) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fmease commented on src/librustdoc/clean/mod.rs:
```diff
@@ -955,6 +956,43 @@ fn clean_ty_generics<'tcx>(
     }
 }
 
+fn clean_ty_typealias_inner_type<'tcx>(
```
`typealias` without an `_` is not that great, what about just

```suggestion
fn clean_ty_alias_inner_type<'tcx>(
```
```

jsha commented:
This is a great idea. It took me a minute to understand it, so I'm paraphrasing to make sure I got it right.

It's often the case that the types of a struct's fields depend on a type parameter of that struct. For instance:

```rust
struct Foo<T> {
  bar: T
}
```

With larger structs this can get quite complicated. Crate authors will sometimes use a type alias to simplify usage of a common set of type parameters. For instance:

```rust
type FooU8 = Foo<u8>;
```

Right now, the doc page for `FooU8` doesn't show very much. It only shows the type alias as it is written in the code. However, it would be useful for doc readers to see a "fully expanded" version of the target type, with all type parameters assigned:

```rust
type FooU8 = Foo<u8>;

// Equivalent to:
struct FooU8 {
  bar: u8
};
```

In this PR, the relevant code is triggered only when all type parameters (other than lifetimes) are assigned. Or in other words, when there are no "free variables" among the type parameters. Can you add a little explanation of why? Is it implementation challenges or UX concerns?

The other change in this PR is that such type aliases "hoist" documentation from their target's fields or variants. That is, documentation that appears on the target's fields or variants now also appears on the documentation page for the type alias, with their type parameters fully resolved. For consistency it seems like "hoisting" should happen even when the target type has no type parameters. Is there something I'm missing?

@rfcbot concern consistency

There are two main things I'd like tweaked. We should avoid using the `<details>` interaction for this. We already rely on that interaction too much, and it creates layout problems and also allows us to dodge questions about the relative importance of different sections. When we show it, the pseudo-declaration is important. Therefore it should get its own h2-level heading, and appear right below the main documentation block. If users want to skip past it to the next section heading, that's what the sidebar links are for.

@rfcbot concern section-heading

The other tweak I'd like to request: this behavior should be documented in the rustdoc book. In particular we should explain the rules for when it's triggered (and add a comment in the implementation saying where to update the rustdoc book if the implementation changes).

@rfcbot concern documentation

Thanks again for working on this. I think it has the potential to untangle complicated types in a lot of cases. Oh, speaking of which: The motivation for this PR is discussed mainly around internal compiler types, but I think it will also be useful in public crates. Would you mind grabbing a couple of examples of popular crates that use type aliases and showing before/after screenshots?

Urgau commented:
> This is a great idea. It took me a minute to understand it[..]

Sorry for this, I tried making it as clear as possible.

> In this PR, the relevant code is triggered only when all type parameters (other than lifetimes) are assigned. Or in other words, when there are no "free variables" among the type parameters. Can you add a little explanation of why? Is it implementation challenges or UX concerns?

This isn't an implementation challenge, rather it's a usefulness issue. I don't know if it's very useful to show the type definition for a "partial" type, but this could lifted pretty easily.

> The other change in this PR is that such type aliases "hoist" documentation from their target's fields or variants. That is, documentation that appears on the target's fields or variants now also appears on the documentation page for the type alias, with their type parameters fully resolved. For consistency it seems like "hoisting" should happen even when the target type has no type parameters. Is there something I'm missing?

I don't think it is very useful to duplicate the type definition when the types aren't different. I think it could create more confusion than be helpful, in part because we don't show the associated methods for the "inner type" in the type alias page, only the ones that are implemented onto the type alias.

> Thanks again for working on this. I think it has the potential to untangle complicated types in a lot of cases. Oh, speaking of which: The motivation for this PR is discussed mainly around internal compiler types, but I think it will also be useful in public crates. Would you mind grabbing a couple of examples of popular crates that use type aliases and showing before/after screenshots?

Sure, do you have a crate in mind ?

jsha commented:
> > relevant code is triggered only when all type parameters (other than lifetimes) are assigned
> This isn't an implementation challenge, rather it's a usefulness issue. I don't know if it's very useful to show the type definition for a "partial" type, but this could lifted pretty easily.

I think the most important thing is for doc readers and doc authors to have a simple and straightforward way to understand what rustdoc does as a tool. So we should choose a simple rule (type aliases always "hoist" fields / variants; type aliases always show their "expanded type") over a complex one, even if that means the results are not 100% useful.

> Sure, do you have a crate in mind ?

Nothing in mind; I was thinking it would make sense grab a few of the "top crates" used by e.g. the Playground, and click through their docs looking for a "Type Aliases" (or "Typedefs") section from the top-level index page.

Urgau commented:
> Also, what does https://doc.rust-lang.org/nightly/std/io/type.Result.html look like with this patch?

With the current, it looks the same. Nothing changed, and I think that a good thing and the "inner type" definition wouldn't be very useful here.

> > > relevant code is triggered only when all type parameters (other than lifetimes) are assigned
> > > This isn't an implementation challenge, rather it's a usefulness issue. I don't know if it's very useful to show the type definition for a "partial" type, but this could lifted pretty easily.

Well the rule is IMO pretty simple: does the type alias point to a concrete type (ie no generics module lifetimes).

Urgau committed:
rustdoc: remove details for type alias inner type and fix sidebar

Urgau committed:
rustdoc: always print type alias inner type (with it's where clauses)

Urgau commented:
@jsha I've pushed 2 new commits, the first one changes the details to an h2 as requested, can you confirm that it resolve your concern?

The second commit removes all conditions regarding displaying the aliased type section. This should in theory resolve your concern, but I still think some filtering might be good, in particular regarding the last 2 examples. What do think about this?

As for your third concern, I don't know if it still hold with the filtering being gone but if it's the case I would really appreciate some guidance for where to put said documentation since I don't know where to put it in the rustdoc book (it not currently well suited for that).

### Examples from top 50 downloaded crates

*Note: only a very small percentage of crates use type alias in API, here are the example I could find*

Here is an example from `syn` type alias `Result` (de-duplicated from multiple crates):
![image](https://github.com/rust-lang/rust/assets/3616612/47bccb03-a74b-4208-a1f2-60a506894474)

Here is an example from `hashbrown`:
![image](https://github.com/rust-lang/rust/assets/3616612/8f313721-7751-4291-a651-bf9e59afb2de)

Here is example from `crypto_common`:
![image](https://github.com/rust-lang/rust/assets/3616612/941c5591-881d-493a-85b6-7bb0a6926d83)

fmease commented:
I'm noticing that the *Aliased Type* of `DefaultHashBuilder` and the one of `ParBlocks` is not super useful since the all the fields are private / hidden. Should we be stricter for these cases and not show the aliased type at all if its fields were stripped?

jsha commented:
That was my first thought as well. Seeing those seemingly information-low
sections made me second-guess my "we should do the same thing for all case"
proposal.

However, on second reflection these are not information-free. They carry
the quite useful information that the target type is opaque. Without the
new "Aliased type" section, that information is only available on the
target type's page.

@urgau good question on where to put it in the book. When I'm back in front
of a computer I'll take a look and reply. Thanks for the updates!

jsha commented:
@urgau Reviewing the screenshots, the current version looks good to me.

I think https://doc.rust-lang.org/nightly/rustdoc/how-to-read-rustdoc.html is the right page to describe the "Aliased Type" section heading on. I propose that we add a new section of that page (or perhaps a new page nested under it), called "Sections", which can document information about various sections that show up in rustdoc output (e.g. "Implementations", "Trait Implementations", "Implementors", "Auto Trait Implementations", etc). We don't need to fill it with all possible sections to start - describing just the new "Aliased Type" section is fine.

I'll take a stab at writing what I think should go in the rustdoc book documentation:

"A type alias is expanded at compile time to its [aliased type](https://doc.rust-lang.org/reference/items/type-aliases.html). That may involve substituting some or all of the type parameters in the target type with types provided by the type alias definition. The Aliased Type section shows the result of this expansion, including the types of public fields or variants, which may depend on those substitutions."

We should also update https://doc.rust-lang.org/reference/items/type-aliases.html so that it mentions (and preferably defines) the term "aliased type."

@rfcbot resolve section-heading
@rfcbot resolve consistency

It's possible to define a public type alias to a private target type. Currently, the target type isn't rendered in rustdoc at all. With this change, I think we would be rendering some information about that private target type. However, "[private type in public interface](https://github.com/rust-lang/rust/issues/34537)" is a warning and eventually to be an error, so I'm okay with this.

Also, in reading https://doc.rust-lang.org/reference/items/type-aliases.html I learned something - in addition to substituting types, a type alias can impose new trait bounds. Below is some sample code including trait bounds and a couple of other cases that I think would be useful in unittests.

```rust
#![feature(lazy_type_alias)]
#![allow(incomplete_features)]
#![allow(dead_code)]

pub struct Pair<A, B> {
  first: A,
  second: B,
}

pub type ReversedTypesPair<Q, R> = Pair<R, Q>;

pub type ReadWrite<R, W> = Pair<R, W> where R: std::io::Read, W: std::io::Write;

pub type VecPair<U, V> = Pair<Vec<U>, Vec<V>>;

fn main() {}
```

Urgau committed:
rustdoc: start new "Sections" section in the book with Aliased Type

fmease commented on tests/rustdoc/typedef-inner-variants.rs:
```diff
@@ -0,0 +1,146 @@
+// @count - '//*[@id="variants"]' 0
+// @count - '//*[@id="fields"]' 1
+// @count - '//span[@class="where fmt-newline"]' 2
+pub type ReadWrite<R, W> where R: std::io::Read, W: std::io::Write = Pair<R, W>;
```
I'm torn on this specific test. This where-clause is basically meaningless since you haven't placed `#![feature(lazy_type_alias)]` at the top making rustc essentially ignore it. Indeed, we can test for it and render it since it's returned by `tcx.generics_of` but it's misleading in my opinion. This triggers the lint warning `type_alias_bounds` with rustc but not with rustdoc since it doesn't run every analysis step.

However, I don't want you enable lazy type aliases for the entire test file since it's an *incomplete* and highly semantics-altering feature. I'd prefer you move this specific test to a separate file.

NB: If you enable lazy type aliases, you are forced to move the where-clause to a trailing position and you will probably find out that rustdoc still renders it in a leading position. Don't worry about that, I've been meaning to fix that for while.

```suggestion
pub type ReadWrite<R, W> = Pair<R, W>
where
    R: std::io::Read,
    W: std::io::Write;
```

CC @jsha

Urgau replied:
Yeah, I explicitly wanted to avoid having `#![feature(lazy_type_alias)]`, which is why I removed it thinking it would be equivalent, to bad that it isn't.

Changed to a separate test as requested.

fmease replied:
Makes sense, thanks a lot! Yeah, the feature is only meant to be stabilized as part of a new edition since it's a breaking change. I've worked a bit on it and I've nominated it for stabilization in Rust 202X.

I appreciate how quickly you address review suggestions. Sorry for having so many haha ^^

fmease reviewed (changes requested):
<details><summary>Outdated</summary>

At the moment, you copy the generics from the type alias to the inner type but discard the generics of the inner type itself. Consider:

```rs
#![feature(lazy_type_alias)]
pub struct S<T: Copy> { pub f: T }
pub type A<X: Eq> = S<X>;
```

It renders `X: Eq` instead of ideally `X: Copy + Eq`:

```diff
  Aliased Type

- struct A<X: Eq> {
+ struct A<X: Copy + Eq> {
      pub f: X,
  }
```

This can be accomplished by merging the generics after having "instantiated" (substituted) the generics of the inner type first. Not sure how easy that would be. I'm fine with doing it in a follow-up PR (you or me).

</details>

I do have one suggestion that I'd like to see addressed before this PR gets merged.

Urgau commented:
> At the moment, you copy the generics from the type alias to the inner type but discard the generics of the inner type itself. Consider:
> 
> ```rust
> #![feature(lazy_type_alias)]
> pub struct S<T: Copy> { pub f: T }
> pub type A<X: Eq> = S<X>;
> ```

I wouldn't expect this to compile due to the missing `Copy` bound on `X`, and indeed it deosn't compile! Did you mean something else ?
```
error[[E0277]](https://doc.rust-lang.org/nightly/error_codes/E0277.html): the trait bound `X: Copy` is not satisfied
 --> src/main.rs:6:21
  |
6 | pub type A<X: Eq> = S<X>;
  |                     ^^^^ the trait `Copy` is not implemented for `X`
  |
note: required by a bound in `S`
 --> src/main.rs:5:17
  |
5 | pub struct S<T: Copy> { pub f: T }
  |                 ^^^^ required by this bound in `S`
help: consider further restricting this bound
  |
6 | pub type A<X: Eq + std::marker::Copy> = S<X>;
  |                  +++++++++++++++++++
```

fmease commented:
Ah, you are absolutely right! I totally forgot about the semantics of lazy type aliases there :sweat_smile: Good catch! Right, no need to merge the generics, since rustc already ensures that user does that!

jsha commented:
Excellent, thanks.

@rfcbot resolve documentation

I'm approving for the UX aspects; I haven't reviewed the implementation.

GuillaumeGomez commented:
Thanks everyone!

@bors r+

bors commented:
:pushpin: Commit 85e4b89cf615c112c72243a2d8f4a55010ecbd4b has been approved by `GuillaumeGomez`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=GuillaumeGomez 85e4b89cf615c112c72243a2d8f4a55010ecbd4b -->
<!-- homu: {"type":"Approved","sha":"85e4b89cf615c112c72243a2d8f4a55010ecbd4b","approver":"GuillaumeGomez","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 85e4b89cf615c112c72243a2d8f4a55010ecbd4b with merge 70c7e4d21c7249af4222a10d09dc03286aa1e787...
<!-- homu: {"type":"BuildStarted","head_sha":"85e4b89cf615c112c72243a2d8f4a55010ecbd4b","merge_sha":"70c7e4d21c7249af4222a10d09dc03286aa1e787"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6112312903/job/16592381668)
Approved by: GuillaumeGomez
Pushing 70c7e4d21c7249af4222a10d09dc03286aa1e787 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"GuillaumeGomez","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6112312903/job/16592381668"},"merge_sha":"70c7e4d21c7249af4222a10d09dc03286aa1e787"} -->

rust-timer commented:
Finished benchmarking commit (70c7e4d21c7249af4222a10d09dc03286aa1e787): [comparison URL](https://perf.rust-lang.org/compare.html?start=f06b7c59a383d51eee35470fd9cd1e0e545b5a5e&end=70c7e4d21c7249af4222a10d09dc03286aa1e787&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=f06b7c59a383d51eee35470fd9cd1e0e545b5a5e&end=70c7e4d21c7249af4222a10d09dc03286aa1e787&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5% | [0.5%, 0.5%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.5% | [0.5%, 0.5%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=f06b7c59a383d51eee35470fd9cd1e0e545b5a5e&end=70c7e4d21c7249af4222a10d09dc03286aa1e787&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.3%  | [1.3%, 1.3%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.6% | [-3.6%, -3.6%] | 1     |
| All ❌✅ (primary)                 | 1.3%  | [1.3%, 1.3%]   | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=f06b7c59a383d51eee35470fd9cd1e0e545b5a5e&end=70c7e4d21c7249af4222a10d09dc03286aa1e787&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=f06b7c59a383d51eee35470fd9cd1e0e545b5a5e&end=70c7e4d21c7249af4222a10d09dc03286aa1e787&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 629.026s -> 628.819s (-0.03%)
**Artifact size**: 318.02 MiB -> 318.09 MiB (0.02%)

jplatte commented:
Wow, this is amazing! Thanks to everybody involved in landing this 🤩
<sub>Could it be extended to also show methods on type aliases? 🙏🏼</sub>

Urgau commented:
> Wow, this is amazing! Thanks to everybody involved in landing this 🤩 Could it be extended to also show applicable methods on type aliases? 🙏🏼

Currently working on it!

GuillaumeGomez commented:
There was https://github.com/rust-lang/rust/pull/112429 which we needed to revert. But luckily we now have https://github.com/rust-lang/rust/pull/115201 which should be merged soon.
